subsys/bluetooth/host/gatt.c

```diff
@@ -742,7 +742,6 @@ void bt_gatt_init(void)
 #if defined(CONFIG_BT_GATT_DYNAMIC_DB)
 #if defined(CONFIG_BT_GATT_CACHING)
 	k_delayed_work_init(&db_hash_work, db_hash_process);
-	db_hash_gen(false);
```

Comment:
Afaik this is called here since the internal services, actually any static services, won't call register so it wont generate the initial hash by itself.

Reply:
Makes sense, db_hash_process is only initialized here. So if all services are static services then db_hash_read will read the value generated here.

Reply:
If all services are static services and dynamic_db is disabled, this would not have been called anyway as it is under CONFIG_BT_GATT_DYNAMIC_DB.

Reply:
But I can also fix this bug (db_hash_gen not being run when CONFIG_BT_GATT_DYNAMIC_DB is unset) in this same PR.

Reply:
@fnde-ot Do you have the traces when the hash is generated too many times, afaik the db_hash_gen would actually cancel the work if in progress or perhaps you are trying to make it run only once when all the dynamic services are registered as well? We could also defer in the init and postpone the first hash to be done after the system init, or alternatively, we could have a precalculate hash as Kconfig option so we could skip the entire hash generation on init, with that we could default to static services defined by the stack but the application could override it if they had other static services registered as well.

Reply:
For the static-only case the precalculated hash idea is a great idea, but it is not what I am trying to fix in this PR.

From what I gather, the current idea is to add a delayed (10ms) job to run db_hash_gen in a work thread when registering a service, so if you register all your services within those 10ms the job will only be run once.

The bug is that we also run db_hash_gen right after setting up the delayed job.

Reply:
BTW, I will try to get a trace to confirm!

Reply:
@fnde-ot What I meant is that DYNAMIC_DB is set, but `bt_gatt_service_register` is never called (or not called before `db_hash_read`). In which case `k_delayed_work_remaining_get` would return zero, and the db_hash value is returned without having been calculated.

So maybe the solution is to submit the db_hash work in `bt_gatt_init`?

Reply:
Yep, Id just submit the work within bt_gatt_init since that perhaps is being called from within bt_gatt_service_register so the hash will immediately change anyway, also if there are more services registered that way we will keep delaying until all services are registered.

Reply:
Here is the code an the shell doing gatt register before bt init to force bt_gatt_service_register to run bt_gatt_init:

https://gist.github.com/Vudentz/4cb49802d44fbd0d6c2ffaccb11b0a88

Reply:
We may want to make DB_HASH_TIMEOUT configurable though since 10 ms may not be enough in certain cases causing the hash to be generated multiple times.

Reply:
Agreed, my mistake: I had misread the code and thought the job was submitted in bt_gatt_init, however it was only a k_delayed_work_init and not submit.

Reply:
Should it also be run when CONFIG_BT_GATT_DYNAMIC_DB is unset (only static services)?

Reply:
BT_GATT_CACHING actually depends on BT_GATT_DYNAMIC_DB:

```
config BT_GATT_CACHING
	bool "GATT Caching support"
	default y
	depends on BT_GATT_DYNAMIC_DB
```

So one cannot enable GATT (Robust) Caching without enabling the dynamic DB, since the former requires the Service Changed characteristic in order to handle change aware state, though perhaps it is not invalid to expose the hash even if it cannot be changed, in which case it would be better to precalculate the hash at build time but that is probably a topic to a different PR.

Reply:
So we only support Service Changed when changes occur at runtime? What about when we update the code on a device, can't it also support Service Changed (if all services were to be static)?

Reply:
> So we only support Service Changed when changes occur at runtime? What about when we update the code on a device, can't it also support Service Changed (if all services were to be static)?

If by update the code you meant flash the device then it must support Service Changed to begin with which means any device which is capable of flashing/dfu shall have dynamic DB enabled, or it must guarantee the database never changes after an update, otherwise a remote device won't be able to detect the changes when in reconnects.

Reply:
Yes that's what I meant, so it means we must enable dynamic DB even if all our services are static and we don't use the register_service functions. Good to know! :)

Is the updated PR good enough? (I think maybe the hash pre-calculation at build time should be done in another PR)

Reply:
Very good, thank you. I don't think the pre-calculation is needed.

Reply:
Thank you guys for the review :)